electrumsv/network.py

```diff
@@ -1022,305 +291,199 @@ async def _main_task(self) -> None:
+        selection_candidate = random.choice(servers) if servers else None
+        # NOTE: Need to add to self.chosen_servers set here because doing it after an
+        # `await _random_server()` opens the possibility for duplicate selections
+        if selection_candidate:
+            self.chosen_servers.add(selection_candidate)
+        return selection_candidate
+
+    async def _random_server(self, retry_timeout: int=10) -> SelectionCandidate:
         while True:
-            server = self._random_server_nowait(protocol)
+            server = self._random_server_nowait()
             if server:
                 return server
-            await sleep(10)
+            await sleep(retry_timeout)
 
-    async def _main_session(self) -> SVSession:
-        while True:
-            session = self.main_session()
-            if session:
-                return session
-            await self.sessions_changed_event.wait()
-
-    async def _random_session(self) -> SVSession:
-        while not self.sessions:
-            logger.info('waiting for new session')
-            await self.sessions_changed_event.wait()
-        return random.choice(self.sessions)
-
-    #
-    # API exposed to SVSession
-    #
+    async def _maintain_connections(self) -> None:
+        count = 1 if app_state.config.get('oneserver') else 100
 
-    async def session_established(self, session: SVSession) -> bool:
-        self.sessions.append(session)
-        self.sessions_changed_event.set()
-        self.sessions_changed_event.clear()
-        self.trigger_callback(NetworkEventNames.SESSIONS)
-        if session.server is self.main_server:
-            self.trigger_callback(NetworkEventNames.GENERIC_STATUS)
-            return True
-        return False
+        async with TaskGroup() as group:
+            for n in range(0, count):
+                await group.spawn(self._maintain_connection, n)
 
-    async def session_closed(self, session: SVSession) -> None:
-        self.sessions.remove(session)
-        self.sessions_changed_event.set()
-        self.sessions_changed_event.clear()
-        if session.server is self.main_server:
-            self.trigger_callback(NetworkEventNames.GENERIC_STATUS)
-        self.trigger_callback(NetworkEventNames.SESSIONS)
+    async def _maintain_connection(self, n: int) -> None:
+        try:
+            while True:
+                selection_candidate = await self._random_server(retry_timeout=30)
+                try:
+                    await self._monitor_chain_tip_task_async(selection_candidate)
+                except ServiceUnavailableError as e:
+                    # assert selection_candidate.api_server is not None
+                    # logger.debug(f"Server unavailable: %s", selection_candidate.api_server.url)
+                    pass
+                finally:
+                    self.chosen_servers.remove(selection_candidate)
+                await asyncio.sleep(20)
+        except Exception:
```

Comment:
Okay have removed this in my local branch but I think the reason this was put in is because long-running async background tasks swallow exceptions if there is an unexpected exception that we have neglected to catch. Seems to be similar to threading.Thread where debugging can be a nightmare if an unexpected exception gets swallowed.

Reply:
If there are unexpected exceptions either we do not understand what we are doing or we have not tested it enough, best not to catch it and deal with the problems. We can't do anything about threads, but there are many ways we can solve swallowed exceptions in tasks that do not involve this.